File: python/agentpin/jwt.py
"""ES256-only JWT encode/decode/verify for AgentPin."""

import base64
import json

from cryptography.hazmat.primitives import hashes, serialization
from cryptography.hazmat.primitives.asymmetric.ec import ECDSA, EllipticCurvePrivateKey, EllipticCurvePublicKey

REQUIRED_ALG = "ES256"
REQUIRED_TYP = "agentpin-credential+jwt"


def base64url_encode(data: bytes) -> str:
    """Base64url encode bytes (no padding)."""
    return base64.urlsafe_b64encode(data).rstrip(b"=").decode("utf-8")


def base64url_decode(s: str) -> bytes:
    """Base64url decode a string."""
    padding = 4 - len(s) % 4
    if padding != 4:
        s += "=" * padding
    return base64.urlsafe_b64decode(s)


def encode_jwt(header: dict, payload: dict, private_key_pem: str) -> str:
    """Encode a JWT from header + payload + PEM private key."""
    header_json = json.dumps(header, separators=(",", ":"))
    payload_json = json.dumps(payload, separators=(",", ":"))

    header_b64 = base64url_encode(header_json.encode("utf-8"))
    payload_b64 = base64url_encode(payload_json.encode("utf-8"))

    signing_input = f"{header_b64}.{payload_b64}"

    private_key = serialization.load_pem_private_key(private_key_pem.encode("utf-8"), password=None)
    assert isinstance(private_key, EllipticCurvePrivateKey)
    signature = private_key.sign(signing_input.encode("utf-8"), ECDSA(hashes.SHA256()))
    sig_b64 = base64url_encode(signature)

    return f"{signing_input}.{sig_b64}"


def decode_jwt_unverified(jwt_str: str) -> tuple:
    """Decode a JWT without verifying the signature.

    Returns:
        Tuple of (header: dict, payload: dict, signature_b64: str)

    Raises:
        ValueError: if JWT is malformed, wrong algorithm, or wrong type
    """
    parts = jwt_str.split(".")
    if len(parts) != 3:
        raise ValueError("JWT must have 3 parts")

    header = json.loads(base64url_decode(parts[0]))
    payload = json.loads(base64url_decode(parts[1]))

    if header.get("alg") != REQUIRED_ALG:
        raise ValueError(f"Algorithm '{header.get('alg')}' rejected, must be '{REQUIRED_ALG}'")

    if header.get("typ") != REQUIRED_TYP:
        raise ValueError(f"Token type '{header.get('typ')}' rejected, must be '{REQUIRED_TYP}'")

    return header, payload, parts[2]


def verify_jwt(jwt_str: str, public_key_pem: str) -> tuple:
    """Verify a JWT signature using a PEM public key.

    Returns:
        Tuple of (header: dict, payload: dict)

    Raises:
        ValueError: if JWT is invalid or signature verification fails
    """
    header, payload, _sig_b64 = decode_jwt_unverified(jwt_str)

    parts = jwt_str.split(".")
    signing_input = f"{parts[0]}.{parts[1]}"
    sig_bytes = base64url_decode(parts[2])

    public_key = serialization.load_pem_public_key(public_key_pem.encode("utf-8"))
    assert isinstance(public_key, EllipticCurvePublicKey)

    try:
        public_key.verify(sig_bytes, signing_input.encode("utf-8"), ECDSA(hashes.SHA256()))
    except Exception as e:
        raise ValueError(f"JWT signature verification failed: {e}") from e

    return header, payload

File: javascript/src/jwt.js
/**
 * ES256-only JWT encode/decode/verify for AgentPin.
 * No external JWT library — algorithm validation enforced inline.
 */

import { createSign, createVerify, createPrivateKey, createPublicKey } from 'crypto';

const REQUIRED_ALG = 'ES256';
const REQUIRED_TYP = 'agentpin-credential+jwt';

/**
 * Base64url encode bytes (no padding).
 * @param {Buffer|Uint8Array} data
 * @returns {string}
 */
export function base64urlEncode(data) {
    return Buffer.from(data)
        .toString('base64')
        .replace(/\+/g, '-')
        .replace(/\//g, '_')
        .replace(/=+$/, '');
}

/**
 * Base64url decode a string.
 * @param {string} str
 * @returns {Buffer}
 */
export function base64urlDecode(str) {
    let base64 = str.replace(/-/g, '+').replace(/_/g, '/');
    const pad = base64.length % 4;
    if (pad === 2) base64 += '==';
    else if (pad === 3) base64 += '=';
    return Buffer.from(base64, 'base64');
}

/**
 * Encode a JWT from header + payload + PEM private key.
 * @param {{ alg: string, typ: string, kid: string }} header
 * @param {object} payload
 * @param {string} privateKeyPem
 * @returns {string} Compact JWT string
 */
export function encodeJwt(header, payload, privateKeyPem) {
    const headerJson = JSON.stringify(header);
    const payloadJson = JSON.stringify(payload);

    const headerB64 = base64urlEncode(Buffer.from(headerJson));
    const payloadB64 = base64urlEncode(Buffer.from(payloadJson));

    const signingInput = `${headerB64}.${payloadB64}`;

    const sign = createSign('SHA256');
    sign.update(signingInput);
    const keyObject = createPrivateKey(privateKeyPem);
    const signature = sign.sign({ key: keyObject, dsaEncoding: 'der' });
    const sigB64 = base64urlEncode(signature);

    return `${signingInput}.${sigB64}`;
}

/**
 * Decode a JWT without verifying the signature.
 * Validates algorithm and token type.
 * @param {string} jwt
 * @returns {{ header: object, payload: object, signatureB64: string }}
 */
export function decodeJwtUnverified(jwt) {
    const parts = jwt.split('.');
    if (parts.length !== 3) {
        throw new Error('JWT must have 3 parts');
    }

    const header = JSON.parse(base64urlDecode(parts[0]).toString());
    const payload = JSON.parse(base64urlDecode(parts[1]).toString());

    if (header.alg !== REQUIRED_ALG) {
        throw new Error(`Algorithm '${header.alg}' rejected, must be '${REQUIRED_ALG}'`);
    }

    if (header.typ !== REQUIRED_TYP) {
        throw new Error(`Token type '${header.typ}' rejected, must be '${REQUIRED_TYP}'`);
    }

    return { header, payload, signatureB64: parts[2] };
}

/**
 * Verify a JWT signature using a PEM public key.
 * @param {string} jwt
 * @param {string} publicKeyPem
 * @returns {{ header: object, payload: object }}
 */
export function verifyJwt(jwt, publicKeyPem) {
    const { header, payload } = decodeJwtUnverified(jwt);

    const parts = jwt.split('.');
    const signingInput = `${parts[0]}.${parts[1]}`;
    const sigBytes = base64urlDecode(parts[2]);

    const verify = createVerify('SHA256');
    verify.update(signingInput);
    const keyObject = createPublicKey(publicKeyPem);
    const valid = verify.verify({ key: keyObject, dsaEncoding: 'der' }, sigBytes);

    if (!valid) {
        throw new Error('JWT signature verification failed');
    }

    return { header, payload };
}
